Detect cold CallTarget invalidation and reset its profile

JohnTortugo opened the pull request:
Closes: #11045 

Please review this patch that modifies Truffle to: 

1. Reset a CallTarget's profile when the `nmethod` associated with it was invalidated by HotSpot because the Code Cache's method flushing heuristics deemed the nmethod to be _cold_.

1. Limit the number of compilations that a CallTarget can have within a time period. This doesn't change the current behavior, instead it just makes the constraint more flexible. I added a new parameter, `MaximumCompilationsWindow`, that the user can set to specify a time window within which the number of compilations will be limited (currently the window is whole duration of application execution).

Tests: 
 - OSX AArch64 with LabsJDK "tip" using "mx gate"
 - Linux x65 with LabsJDK "tip" using "mx gate"
 - GitHub Actions
- We have been running an internal system with _similar_ changes for months and we didn't notice any regression.

JohnTortugo commented:
Can someone please take a look? /cc @chumer @dougxc ?

JohnTortugo commented on truffle/src/com.oracle.truffle.runtime/src/com/oracle/truffle/runtime/hotspot/HotSpotOptimizedCallTarget.java:
```diff
@@ -195,4 +213,51 @@ public SpeculationLog getCompilationSpeculationLog() {
+        runtime().getListener().onCompilationDeoptimized(this, frame, getInvalidationReasonDescription());
+    }
+
+    @TruffleBoundary
```
I had to make this a TruffleBoundary because an unit test was failing because a compilation was getting too big due to recursive inlining during PE. Since the test was triggering that I suspected that it could eventually affect an end user as well.

JohnTortugo commented on truffle/src/com.oracle.truffle.runtime/src/com/oracle/truffle/runtime/hotspot/HotSpotOptimizedCallTarget.java:
```diff
@@ -174,9 +178,23 @@ private void tetherSpeculationLog(HotSpotNmethod nmethod) throws Error, Internal
+        boolean isValid = installedCode.isValid();
+        if (!isValid && installedCode != INVALID_CODE) {
+            if (getInvalidationReason() == ((HotSpotTruffleRuntime) runtime()).getColdMethodInvalidationReason()) {
+                invalidateExistingCode();
```
> How do we know isValid() is called often enough?

Do you mean often enough so that we detect "soon" that the call target was invalidated? My understanding is that there is no requirement for how "soon" we detect the invalidation. We just need to eventually detect it and since the `isValid` method is always called before a new compilation I decided to implement things like this.

> ... listeners could call into isValid() too!

The existing code is invalidated before `onCompilationDeoptimized` is called, which prevents a recursion.

> Can we call this maybe between compilations on compiler threads?

I'm not sure what's your suggestion here. Do you mean iterate on all current compiled call targets to see if any was invalidated?

> Can't say I am a fan of this solution.

Please, let me know your suggestions!

JohnTortugo replied:
I need to reset the CT (Call Target) profile when its INSTALLED_CODE was invalidated by HotSpot because it was cold. This needs to happen before we decide to compile the CT again. Realistically speaking this kind of invalidation can only happen between executions, not during an execution.

JVMCI is not able to call a method in Truffle or GraalJIT. Therefore, it cannot
downcall to notify that the CT was invalidated. Thus, the identification that
the CT was invalidated is a passive event that happens in Truffle.

Approaches that we can use to detect that the CT was invalidated:
- Check before every execution of the CT in interpreter mode; before we check if the CT should be compiled.
  * Pros:
    - Relatively simple to implement.
    - This is a direct way to achieve what we want: prevent recompilation of cold method.
  * Cons:
    - Executed every time that we want to execute the CT.

- Use a separate thread to detect all invalidations.
  * Blockers:
    - Will require maintaining a list of all compiled non-invalidated CT.

- When we are installing a new code.
  * Blockers:
    - This is too late. There is already a new compilation of the CT.

chumer commented on truffle/src/com.oracle.truffle.runtime/src/com/oracle/truffle/runtime/OptimizedCallTarget.java:
```diff
@@ -537,6 +540,7 @@ public final RootNode getRootNode() {
     public final void resetCompilationProfile() {
         this.callCount = 0;
         this.callAndLoopCount = 0;
+        this.timeOfFirstCompilationInWindow = Instant.now();
```
So because the compilation queue is flushed often, you want to limit the maximum compilations per certain time frame? Effectively defeating the maximum compilations flag. Wouldn't it be better to just reset the compilation count on a flush? Or do I miss something here?

JohnTortugo commented:
@chumer, I did a refactoring of the code following your suggestions. Basically, I changed two things:

- I dropped the changes related to resetting number of compilations based on time. The approach now is to reset `successfullRecompilationCount` when the method was invalidated because it was cold.
- I reverted the changes in `HotspotOptimizedCallTarget` to detect that the method was invalidated. I implemented the idea that you suggested: verifying the invalidation reason of the call target when it's about to be recompiled.

JohnTortugo commented:
Gentle ping. @chumer - Please, take a look when you can.

JohnTortugo commented on truffle/src/com.oracle.truffle.runtime/src/com/oracle/truffle/runtime/OptimizedTruffleRuntime.java:
```diff
@@ -1491,4 +1496,15 @@ public enum CompilationActivityMode {
+     * @param callTarget - The OptimizedCallTarget for the compilation.
+     * @return true if the compilation should be aborted.
+     */
+    protected boolean shouldAbortCompilation(OptimizedCallTarget callTarget) {
```
I moved the code to `HotSpotOptimizeddCallTarget#prepareForCompilation`.
